### ISSUE TITLE
Unexpected top level exception - multiple dev files

### ISSUE THREAD

**Issue by SMJ93** [kraihn/cordova-plugin-tag-manager#19]
I experienced this problem before with Facebook and Google analytics plugins - the following solution fixed it for me:

http://www.thedevline.com/2014/06/unexpected-top-level-exception.html

I now need to install Google Tag manager in addition to these other plugins. Installing the plugin as it says in the read me triggers the following error when building on Android at 94% (it works fine on iOS):

```
:preDexDebug UP-TO-DATE
:dexDebug

UNEXPECTED TOP-LEVEL EXCEPTION:
com.android.dex.DexException: Multiple dex files define Lcom/google/android/gms/analytics/internal/Command$1;
at com.android.dx.merge.DexMerger.readSortableTypes(DexMerger.java:596)
at com.android.dx.merge.DexMerger.getSortedTypes(DexMerger.java:554)
at com.android.dx.merge.DexMerger.mergeClassDefs(DexMerger.java:535)
at com.android.dx.merge.DexMerger.mergeDexes(DexMerger.java:171)
at com.android.dx.merge.DexMerger.merge(DexMerger.java:189)
at com.android.dx.command.dexer.Main.mergeLibraryDexBuffers(Main.java:502)
at com.android.dx.command.dexer.Main.runMonoDex(Main.java:334)
at com.android.dx.command.dexer.Main.run(Main.java:277)
at com.android.dx.command.dexer.Main.main(Main.java:245)
at com.android.dx.command.Main.main(Main.java:106)

:dexDebug FAILED

FAILURE: Build failed with an exception.
```

Additional info:

```
Error Code:
2
  Output:

    UNEXPECTED TOP-LEVEL EXCEPTION:
    com.android.dex.DexException: Multiple dex files define    Lcom/google/android/gms/analytics/internal/Command$1;
    at com.android.dx.merge.DexMerger.readSortableTypes(DexMerger.java:596)
    at com.android.dx.merge.DexMerger.getSortedTypes(DexMerger.java:554)
    at com.android.dx.merge.DexMerger.mergeClassDefs(DexMerger.java:535)
    at com.android.dx.merge.DexMerger.mergeDexes(DexMerger.java:171)
    at com.android.dx.merge.DexMerger.merge(DexMerger.java:189)
    at com.android.dx.command.dexer.Main.mergeLibraryDexBuffers(Main.java:502)
    at com.android.dx.command.dexer.Main.runMonoDex(Main.java:334)
    at com.android.dx.command.dexer.Main.run(Main.java:277)
    at com.android.dx.command.dexer.Main.main(Main.java:245)
    at com.android.dx.command.Main.main(Main.java:106)



* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output.

BUILD FAILED

Total time: 10.506 secs
```

Is it possible to run Google analytics with Tag manager? If so any idea how to get around this error?

Cheers, Stu

**Comment by kraihn** [comment]
I'm not sure I understand your question. Are you asking if you can run Google Analytics in parallel with Tag Manager? If so, then no. There is no reason to do so because Tag Manager is fully capable of forwarding all analytics information to Google Analytics without a second locally installed plugin in your mobile application.

**Comment by SMJ93** [comment]
Basically I have several plugins in my Cordova app that depends on Google Analytics / Google Mobile Services.

Here is a list of all my plugins (I have highlighted which ones use GA / GMS)

**com.appsflyer.phonegap.plugins.appsflyer 1.0.1 "AppsFlyer"**
**com.google.playservices 22.0.0 "Google Play Services for Android"**
com.philmerrell.cordova.canopen 0.1 "CanOpen"
**com.phonegap.plugins.facebookconnect 0.11.0 "Facebook Connect"**
**com.pushwoosh.plugins.pushwoosh 3.6.5 "Pushwoosh"**
cordova-plugin-device 1.0.1 "Device"
cordova-plugin-dialogs 1.1.1 "Notification"
cordova-plugin-geolocation 1.0.1 "Geolocation"
cordova-plugin-globalization 1.0.1 "Globalization"
**cordova-plugin-google-analytics 0.8.0 "Google Universal Analytics Plugin"**
cordova-plugin-inappbrowser 1.0.1 "InAppBrowser"
cordova-plugin-network-information 1.0.1 "Network Information"
cordova-plugin-splashscreen 2.1.0 "Splashscreen"
cordova-plugin-statusbar 1.0.1 "StatusBar"
cordova-plugin-whitelist 1.0.0 "Whitelist"
nl.x-services.plugins.launchmyapp 3.2.4 "Custom URL scheme"
org.pushandplay.cordova.apprate 1.1.7 "AppRate"

These all work fine together, but adding in Google Plugin Tag manager breaks the build - causes the error above.

Basically different versions of the GA / GMS library are being loaded into the app causing the error.

Any idea how to solve this?

Cheers, Stu

**Comment by kraihn** [comment]
@SMJ93, what is the reason for using both cordova-plugin-google-analytics and cordova-plugin-tag-manager? I looked through the plugins you mentioned, and none of them require cordova-plugin-google-analytics. The two plugins clash because they have the same namespace dependencies, but you only need one plugin.

cordova-plugin-tag-manager supercedes cordova-plugin-google-analytics and makes it obsolete because GTM sends analytics to Google Universal Analytics.

**Comment by SMJ93** [comment]
My client wants to be able to use both.

So if I uninstall the Google Analytics plugin and install this plugin I will be able to get tag manager and GA data?

Thanks for the quick response!

Cheers, Stu

**Comment by kraihn** [comment]
Yes, that's how Google Tag Manager works. The GTM model is that they are a data aggregator, and then based on conditions you set, your analytics get forwarded on to one or multiple analytics services. With Google Analytics being the default option.

I would highly recommend logging into the Tag Manager product. It's free, and should help you understand what it really does.

**Comment by SMJ93** [comment]
Ok thanks @kraihn! I will look into it.
